[MRG + 1] optimize rfecv by eliminating the ```for``` loop

xuewei4d opened the pull request:
Fixes #4382 , I am not sure this is good enough. Any comment is warmly welcome. 

Basically, in this PR, `RFE` has  `_fit` which is almost identical to previous `fit`. `_fit` has a third parameter `step_score` which is a function evaluating the score of a parameter. `step_score` takes `estimator` and `support`, then gives a score. In `RFECV` there is no `for` loop right now. For each fold, one `rfe._fit` is called with the return value of `_step_score_on_test`. `_step_score_on_test` acts like a factory method. It accepts `X_test`, `y_test`, and builds a `step_score` function, which use `_score` just like the implementation in the `for` loop before this PR. By doing so, `RFE` is not aware of the test data. Because I think it is weird that implementing a new `fit` function in RFE which explicitly accepts test data to get the score.

The number of subsets of features needs to be evaluated is `ceil((n_features - n_features_to_select)/step) + 1`. For example `n_features = 10`, `n_features_to_select=1`, `step=2`. The sizes of each subsets of features are `[10, 8, 6, 4, 2, 1]`. In `RFE.fit`'s `for` loop, the subsets which has size of `[10, 8, 6, 4, 2]` is evaluated. The last one is evaluated in the final estimator's `fit` after `for` loop. This explains the size of `scores_` in `RFE`.

xuewei4d committed:
ENH optimize rfecv by eliminating the for loop

amueller commented on sklearn/feature_selection/rfe.py:
```diff
@@ -425,3 +460,8 @@ def fit(self, X, y):
         return self
+
+    @staticmethod
+    def _step_score_on_test(X_test, y_test, scorer):
```
I'm not sure it is worht adding this function, why not just doing the lambda inline?

xuewei4d replied:
Yeah, I agree.

amueller commented on sklearn/feature_selection/rfe.py:
```diff
@@ -146,6 +149,13 @@ def fit(self, X, y):
+
+        if step_score:
+            # Determine the number of subsets of features
+            self.scores_ = np.zeros(
```
wouldn't it be much easier to just create a list and add to that?

xuewei4d replied:
I initially use a list, but I have to get the size the ```scores``` in RFECV beforehand anyway , because ```scores``` in RFECV cannot be a list to do ```scores += rfe.scores_```.  So it does not make much difference, I think.

amueller replied:
sorry, I don't understand this point. It seems you could save these six lines here...

xuewei4d replied:
Oh, I mean. First, ```scores``` in RFECV has to be an array to do ```scores += rfe.scores_``` in L435. It cannot a list in RFECV. If ```rfe.scores_``` is a list, I have to reverse the list and create a new array in L435 in RFECV. Second, ```scores``` RFECV is just the same as in RFE, it needs to compute the size beforehand anyway as well in L417. So I just do it in RFE as well, L155.  I am OK to remove the computation, actually it is not so complicated....

amueller replied:
I am not saying that it is complicated, I'm saying that you could have fewer lines do the same. You have to compute indices again in ``_do_step_score`` actually you could probably eliminate this function entirely if you used a list, right? You can later convert the list to an array.
I'd just like to have as few indirections and as little code as possible, as that will make it easier for someone to follow the code later.

xuewei4d replied:
I see. :)  I will make it simple and clean...

xuewei4d replied:
I removed the score initialization in RFECV as well.

amueller commented:
have you benchmarked to see if there is any performance improvement?

xuewei4d commented:
I benchmarked on the iris data set, but I guess since the number of feature is small, I did not observe the performance improvement. I am trying to do that on a high-dimension data set.

amueller commented:
try mnist with random forests ;) [if you are patient]

xuewei4d commented:
OK. I will try it. 

I don't know why Travis is not happy on Python2, but pass it on Python3.  I found that in Python2 Numpy is 1.6.1. maybe because of that?

amueller commented:
I haven't looked into the error. Maybe an integer / float division issue?

xuewei4d commented:
No clue about the Travis failed checks on Python2. It is said that the score function expects the data array which has feature selected, but it still receives array without any selection. But I did that in [L409](https://github.com/scikit-learn/scikit-learn/pull/4534/files#diff-589a953cf2ee991da252f7da86c1e5b9R409). Besides, the PR passed on my machines (OS X, Ubuntu) with Python2 and latest Numpy, Scipy.

amueller commented:
I was hoping this is #4528, but it is not. also passing on my box.

amueller commented:
can you try a virtualenv with the same numpy as travis?

xuewei4d commented:
passed with
```python
Python 2.7.6 (default, Mar 22 2014, 22:59:56)
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy as sp
>>> sp.__version__
'0.11.0'
>>> import numpy as np
>>> np.__version__
'1.6.2'
```
:-/

amueller commented:
Fails both on ubuntu install and ubuntu/conda.... meh

amueller commented:
I sent your a PR here: https://github.com/xuewei4d/scikit-learn/pull/1 maybe merge that and we could see what is happening?

amueller commented:
I don't understand what's happening, I must be overlooking something.

amueller commented:
Ah, I think I somewhat understand now, but without reproducing it is a bit hard.

xuewei4d commented:
Me too. I found it is a really weird. 

I inserted some line in ```_build_step_score``` in the last commit, just for testing.

I used IPython2 to test the ```test_rfecv``` function in the ```test_rfe.py```. 

```python
%run sklearn/feature_selection/tests/test_rfe.py
from sklearn.feature_selection.tests.test_rfe import *
test_rfecv()
```
There is no error at all.

But when I use nosetests like 
```nosetests -v sklearn/feature_selection/tests/test_rfe.py```
It reports 
```
IndexError: tuple index out of range
-------------------- >> begin captured stdout << ---------------------
(30, 10)
(10,)

--------------------- >> end captured stdout << ----------------------
```

It seems that something is broken.

xuewei4d commented:
I finally found the "bug" which is a result of  0.10.1 < scipy < 0.13 does not support boolean mask for sparse array, [scipy document](https://docs.scipy.org/doc/scipy-0.15.1/reference/release.0.13.0.html), [stackoverflow](http://stackoverflow.com/questions/20080332/slicing-a-scipy-sparse-matrix-using-a-boolean-mask). The odd  scipy treats boolean mask as integer mask. That is just duplicating the 0th or the 1st column of the sparse matrix. 

In the master branch, in [L152](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/feature_selection/rfe.py#L152) the first author of RFE did the right thing, but there is the exact same bug in [L191](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/feature_selection/rfe.py#L191) which is also made by him. The bug exists since the creation of RFE :-|.

I think it would be better to modify .travis configure file so that it could use the latest scipy.

xuewei4d commented on sklearn/feature_selection/rfe.py:
```diff
@@ -181,14 +188,25 @@ def fit(self, X, y):
         if self.estimator_params:
             self.estimator_.set_params(**self.estimator_params)
-        self.estimator_.fit(X[:, support_], y)
+        self.estimator_.fit(X[:, features], y)
```
The bug in L191 caused by scipy behavior change,

amueller replied:
You can use ``safe_mask``, which will only call ``np.where`` if necessary.

xuewei4d commented:
performance benchmark left...

xuewei4d commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -176,7 +178,8 @@ def test_rfecv():
     # same in sparse
     rfecv_sparse = RFECV(estimator=SVC(kernel="linear"), step=1, cv=5)
-    X_sparse = sparse.csr_matrix(X)
+    X_sparse = sparse.csc_matrix(X)
```
According to the previous commits, it will  be better to use csc matrix, but within SVC estimator, ```check_array``` will convert it into csr matrix anyway.

xuewei4d commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -30,7 +30,7 @@ def __init__(self, foo_param=0):
     def fit(self, X, Y):
         assert_true(len(X) == len(Y))
-        self.coef_ = np.ones(X.shape[1], dtype=np.float64)
+        self.coef_ = np.ones((1, X.shape[1]), dtype=np.float64)
```
I once added an assertion for RFECV to test whether the coef_ has the same length on the second dimension as X has. And I think it would be better to change the coef_ into 2 dimensions even in the test case.

xuewei4d commented on sklearn/tests/test_metaestimators.py:
```diff
@@ -59,7 +59,7 @@ def __init__(self, param=1, hidden_method=None):
 
         def fit(self, X, y=None, *args, **kwargs):
-            self.coef_ = np.arange(X.shape[1])
+            self.coef_ = np.arange(X.shape[1]).reshape(1, -1)
```
Same here

amueller commented:
We should definitely not change travis but fix the bug and add an explicit test for it if possible. (I think you just did that). I'm not sure why you have to reshape anything for that, or convert the sparse matrix. This is just a matter of calling ``np.where`` on a mask, right?

amueller commented:
btw, you can just use ``sklearn.utils.safe_mask``

amueller commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -176,7 +178,8 @@ def test_rfecv():
     rfecv_sparse = RFECV(estimator=SVC(kernel="linear"), step=1, cv=5)
-    X_sparse = sparse.csr_matrix(X)
+    X_sparse = sparse.csc_matrix(X)
+    assert_equal(X_sparse.shape, X.shape)
```
My asserts are really useless, feel free to remove them / my commits.

xuewei4d commented:
I am not sure this is a good way to benchmark it.
```python
X, y = make_classification(n_samples=1000, n_features=50, n_informative=5,
                           n_redundant=2, n_repeated=0, n_classes=8,
                           n_clusters_per_class=1, random_state=0)

svc = SVC(kernel="linear")
rfecv = RFECV(estimator=svc, step=1, cv=StratifiedKFold(y, 2),
              scoring='accuracy')
def run():
    # Build a classification task using 3 informative features
    rfecv.fit(X, y)
    print rfecv.n_features_

if __name__  == '__main__':
    import timeit
    print(timeit.timeit("run()", setup="from __main__ import run", number=10))
```

On my machine, this PR gives 77.35 seconds, sklearn 0.16 gives 113.13 seconds.

amueller commented:
I often use IPython's ``%timeit`` or just ``time.time`` (see the benchmark folder).
It can also be nice to use ``cProfile``, possibly together with runsnakerun.

amueller commented:
btw, you don't have to do ``StratifiedKFold(y, 2)`` you can just use ``2``.

amueller commented:
I think this looks good. Feel free to rename to MRG (+1).

amueller commented on sklearn/feature_selection/rfe.py:
```diff
@@ -367,47 +385,40 @@ def fit(self, X, y):
+            scores.append(np.array(rfe.scores_[::-1]).reshape(1, -1))
+        scores = np.sum(np.concatenate(scores, 0), 0)
+        # The index in scores when 'n_features' features is selected
+        n_feature_index = np.ceil((n_features - n_features_to_select) /
```
It might possibly be easier and less prone to bugs to store this? I remember the formula for this was quite tricky.

amueller replied:
Thee this comment: https://github.com/scikit-learn/scikit-learn/pull/3824#discussion_r19851288
Not sure if we actually added a regression test. If not, maybe we should now?

xuewei4d commented:
I redo it in IPython. 

```python
X, y = make_classification(n_samples=2000, n_features=100, n_informative=1, 
                                          n_redundant=1, n_repeated=0, n_classes=2, 
                                          n_clusters_per_class=1, random_state=0)
rfecv = RFECV(estimator=svc, step=1, cv=2, scoring='accuracy')
%timeit rfecv.fit(X, y)
```
This PR gives ```3: 18.2 s per loop```. sklearn 0.16 gives ```25.9 s per loop```

amueller commented:
That looks like a nice improvement. Can you double check the computation of the ``n_feature_index`` in the cases mentioned in the old PR and see if there are regression tests?

xuewei4d commented:
I checked  the n_feature_index.  I also understand the comment #3824. 

The number of iterations in RFECV of master and in RFE of this PR, is 
```1 + (n_features + step - 2) // step ``` [1] which is also equivalent to my computation 
```1 + ceil((n_features - 1) / float(step)) ``` [2].

Formula [1] is a little tricky indeed, formula [2] is easy to understand. The following comment explains these two.

## Case 1. 
X.shape[1] == 11, n_features_to_select = 1, step = 2, 

The sizes of subsets of features are ```[11, 9, 7, 5, 3, 1]```. Imaging you are standing on the number 11, you walk each step by 2 units. The number of steps you need to take to get number 1 is ```(11 -1) / 2```. That means you take the 1st step to get number 9, and take the 2nd step to number 7. So the numbers you will pass by is ```[9, 7, 5, 3, 1]```.  Plus the start point number 11. 

```(n_features - 1) / 2``` is divisible, formula 1 is equivalent to formula 2.
```
1 + (n_features - 1) // step 
= 1 + (n_features - 1) / step 
= 1 + ceil((n_features - 1) / float(step))
```


## Case 2. 
X.shape[1] == 10, n_features_to_select = 1, step = 2, 
The sizes of subsets of features are ```[10, 8, 6, 4, 2, 1]```. Imaging again :).  Starting form number 10, you take 1st step to get number 8, take 2nd step to get number 6. However, when you are at number 2, you cannot walk anymore with 2units per step. Suppose you take 0.5 step, that mean you could take an half step, or 1 units, then you are done. And don't forget to take the start point into account.

In Formula 1,  ```1 + (n_features - 1) // step + 1```, ```(n_features - 1) // step + 1``` means you cannot walk the last half step (integer division), but you can do extra half step( ```+1```). 
Formula 2 is simple, if you ever run into non-integer steps like 4.5 steps, or 6.5 steps, you could just round up the number of steps (using ceiling function) to finish walking.  Therefore
```
1 + (n_features - 1) // step + 1 
= 1 + ceil((n_features - 1) / float(step))
```

## Combination of two cases of formula 1
Two cases of formula 1 are combined into ```1 + (n_features + step - 2) // step```.
Case 1. 
```
1 + (n_features + step - 2) // step 
= 1 + (n_features  - 1 + step - 1) // step 
= 1 + (n_features - 1) / step + (step - 1) // step
= 1 + (n_features - 1) / step
```

Case 2
```
1 + (n_features + step - 2) // step
= 1 + (n_features - 2) // step + 1
```
since ```n_features - 1 ``` is not divisible by step, ```(n_features - 2)``` could be divisible or not. In either case,  ```(n_features - 2) // step = (n_features - 1) // step``` as long as (n_features - 1) is not divisible by step. 


## Explain ```n_feature_index``` 
(maybe should have a better name). ```n_feature_index```  is the max index of the score arrays. It just one less than the number of subsets of features need check.

Suppose the optimal number of selected feature is ```x```, which corresponds the index ```k``` in the array. 
We could simply compute ```n_features - (n_feature_index - k)*step``` to get ```x```. When ```n_features - 1 ``` is not divisible by step, then this value could be 0, since you just walk past the finish line number 1. So I just take ```max``` function to ensure at least 1 feature is selected.

I think formula 2 is easy to understand, but needs more function calls.

amueller commented:
Thanks for the elaborate explanation, seems good to me.
Still, it would be nice to have a regression test. Also, don't worry about a single call to ``ceil`` ;)

xuewei4d commented:
I am adding a regression test for RFECV, 
but for RFE as well ?

amueller commented:
your call.

xuewei4d committed:
Add regression test for the number of subsets of features

xuewei4d commented:
The regression test cases are added.

amueller commented:
Sorry, that is not what I meant. https://en.wikipedia.org/wiki/Regression_testing
I meant a test that checks that the formula for the indices is correct.

xuewei4d commented:
Oops, sorry for misunderstanding. I will fix it.

agramfort commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -141,6 +144,31 @@ def test_rfe():
+
+    clf = LinearRegression()
+    clf.fit(X_r, y)
+    assert_array_almost_equal(X_r, X[:, coef > 0])
```
`coef > 0 ->  coef !=`

as it makes sense to have negative coefs in regression.

agramfort commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -220,6 +248,39 @@ def test_scorer(estimator, X, y):
+def test_rfecv_regression():
+    X, y, coef = make_regression(n_samples=100, n_features=10,
+                                 n_informative=2, coef=True, random_state=17)
+    X_true = X[:, coef > 0]
```
`coef > 0 ->  coef != 0.`

agramfort commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -220,6 +248,39 @@ def test_scorer(estimator, X, y):
+    X_sparse = sparse.csr_matrix(X)
+    rfecv_sparse.fit(X_sparse, y)
+    X_r_sparse = rfecv_sparse.transform(X_sparse)
+    assert_array_equal(X_r_sparse.toarray(), X_true)
```
you should factorize these two tests either with an aux function _test_rfe or with a for loop looping on a regressor and a classifier.

xuewei4d replied:
do you think it is necessary to add test cases using regressor ?

agramfort commented:
besides LGTM

agramfort commented:
ping me when done

xuewei4d commented:
Since I misunderstood regression test for the function of the number of subsets of features as test cases using regressors, I added some test cases of RFE and RFECV with regressor. So I removed these test cases, and added regression test.

I used a ```for``` loop to remove some redundancy in regression test function ```test_number_of_subsets_of_features```. @agramfort

agramfort commented:
LGTM

+1 for merge

@amueller merge if you're happy as travis :)

amueller commented:
LGTM, thanks @xuewei4d
